### PR TITLE
Scim add EntryReference 

### PR DESCRIPTION
**Pull request by ToxicMushroom**
# Change summary
Replaces the uuid references with a struct containing the uuid and name.

Checklist

- [x] This PR contains no AI generated code
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

So far added a test to check the goal behavior

### CODE REVIEW

**Review by Firstyear — COMMENTED** [review]
As we talked about, I think we should return a different type than Option<ScimValue> which represents that more resolution is needed.

**Comment by ToxicMushroom** [comment]
> As we talked about, I think we should return a different type than Option which represents that more resolution is needed.

Well the commit said not to look at it :)

File: proto/src/scim_v1/server.rs [hunk 1]
```diff
@@ -179,6 +179,182 @@ pub struct ScimOAuth2ClaimMap {
...
+    }
+}
+
+impl From<Vec<ScimKeyInternal>> for ScimResolveStatus {
+    fn from(set: Vec<ScimKeyInternal>) -> Self {
+        Self::Resolved(set.into())
+    }
+}
+
+impl ScimResolveStatus {
```
Comment (Firstyear) [inline proto/src/scim_v1/server.rs:340]: block needs to be #![cfg(feature = "test")]

File: proto/src/scim_v1/server.rs [hunk 2]
```diff
@@ -179,6 +179,182 @@ pub struct ScimOAuth2ClaimMap {
...
+    }
+}
+
+impl From<Vec<ScimApiToken>> for ScimResolveStatus {
+    fn from(set: Vec<ScimApiToken>) -> Self {
+        Self::Resolved(set.into())
+    }
+}
+
+impl From<Vec<ScimIntentToken>> for ScimResolveStatus {
```
Comment (Firstyear) [inline proto/src/scim_v1/server.rs:316]: Can't this be

```
impl From<T> for ScimResolveStatus
where T: Into<ScimValueKanidm> {
    fn from(v: T) -> Self {
        Self::Resolved(v.into())
    }
}
```

File: proto/src/scim_v1/server.rs (hunk 1 above)
Comment (Firstyear) [reply proto/src/scim_v1/server.rs:340]: Or alternately, don't define it here - define it in ValueSet behind the cfg(test) as a private impl. We don't want panics in pub interfaces.

**Review by Firstyear — CHANGES_REQUESTED** [review]

File: proto/src/scim_v1/server.rs (hunk 2 above)
Comment (ToxicMushroom) [reply proto/src/scim_v1/server.rs]: yep ty

File: proto/src/scim_v1/server.rs [hunk 3]
```diff
@@ -180,6 +180,21 @@ pub struct ScimOAuth2ClaimMap {
...
+#[derive(Serialize, Debug, Clone, PartialEq, Eq, ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct ScimReference {
+    pub uuid: Uuid,
+    pub value: String,
+}
+
+#[serde_as]
+#[derive(Serialize, Debug, Clone, PartialEq, Eq)]
+pub enum ScimValueIntermediate {
```
Comment (Firstyear) [inline proto/src/scim_v1/server.rs:193]: Why is this in proto? We never release it publicly?

File: server/lib/src/entry.rs [hunk 4]
```diff
@@ -2353,6 +2372,37 @@ impl Entry<EntryReduced, EntryCommitted> {
     }
 }
 
+fn resolve_interim(
```
Comment (Firstyear) [inline server/lib/src/entry.rs:2375]: resolve_scim_interim

File: server/lib/src/entry.rs [hunk 5]
```diff
@@ -29,12 +29,25 @@ pub use std::collections::BTreeSet as Set;
 use std::collections::{BTreeMap as Map, BTreeMap, BTreeSet};
 use std::sync::Arc;
 
+use crate::be::dbentry::{DbEntry, DbEntryVers};
+use crate::be::dbvalue::DbValueSetV2;
+use crate::be::{IdxKey, IdxSlope};
+use crate::credential::apppwd::ApplicationPassword;
+use crate::credential::Credential;
```
Comment (Firstyear) [inline server/lib/src/entry.rs:36]: Was this cargo fmt re-arranging things? Or your editor?

File: server/lib/src/entry.rs [hunk 6]
```diff
@@ -2230,15 +2230,34 @@ impl Entry<EntryReduced, EntryCommitted> {
...
+    ) -> Result<ScimEntryKanidm, OperationError> {
+        let result: Result<BTreeMap<Attribute, ScimValueKanidm>, OperationError> = self
             .attrs
             .iter()
             // We want to skip some attributes as they are already in the header.
             .filter(|(k, _vs)| **k != Attribute::Uuid)
-            .filter_map(|(k, vs)| vs.to_scim_value().map(|scim_value| (k.clone(), scim_value)))
+            .filter_map(|(k, vs)| {
+                let opt_resolve_status = vs.to_scim_value();
+                let res_opt_scim_value = match opt_resolve_status {
```
Comment (Firstyear) [inline server/lib/src/entry.rs:2244]: look at 'transpose()' on Result.

File: server/lib/src/entry.rs [hunk 7]
```diff
@@ -2353,6 +2372,37 @@ impl Entry<EntryReduced, EntryCommitted> {
...
+) -> Result<Option<ScimValueKanidm>, OperationError> {
+    match scim_value_intermediate {
+        ScimValueIntermediate::Refer(uuid) => {
+            if let Some(option) = read_txn.uuid_to_spn(uuid)? {
+                Ok(Some(ScimValueKanidm::EntryReference(ScimReference {
+                    uuid,
+                    value: option.to_proto_string_clone(),
+                })))
+            } else {
+                // TODO: didn't have spn, fallback to uuid.to_string ?
```
Comment (Firstyear) [inline server/lib/src/entry.rs:2387]: Doesn't uuid_to_spn already do this?

**Review by Firstyear — CHANGES_REQUESTED** [review]

File: server/lib/src/entry.rs (hunk 7 above)
Comment (ToxicMushroom) [reply server/lib/src/entry.rs:2387]: Doesn't appear like that according to the test
https://github.com/ToxicMushroom/kanidm/blob/957aaabe5e6cb6a950ee9f8aef63d368e7577241/server/lib/src/server/mod.rs#L2362
      VV
 ```rust
      
        // Name doesn't exist
        let r1 = server_txn.uuid_to_spn(uuid!("bae3f507-e6c3-44ba-ad01-f8ff1083534a"));
        // There is nothing.
        assert_eq!(r1, Ok(None));
   ```

File: server/lib/src/entry.rs (hunk 5 above)
Comment (ToxicMushroom) [reply server/lib/src/entry.rs:36]: Should be cargo fmt's doings.
with toolchain: stable-x86_64-unknown-linux-gnu (default)
rustc 1.81.0

**Review by Firstyear — APPROVED** [review]
Thank you!

**Pull request merged**